In [5]:
import pandas as pd
import numpy as np
from scipy import stats
from scipy.stats import chisquare
df = pd.read_csv('datasets/Ashopping.csv', sep=',', encoding='CP949')

In [6]:
X=pd.crosstab(df.클레임접수여부, df.구매유형, margins=True)
X.head()

구매유형      1    2    3    4   All
클레임접수여부                         
0        24  197   57  263   541
1        19  120   87  233   459
All      43  317  144  496  1000

[구매유형] 1:1회성구매형, 2:실용적구매형, 3: 명품구매형, 4:집중구매형   
[클레임접수여부] 0:접수하지않음, 1:접수경험있음   
   
     
       
#### 적합도 분석
H0(귀무가설)=클레임 접수 고객의 구매유형별 비율은 1회성구매형 10%, 실용적구매 30%,  
명품구매형 20%, 집중구매형 40%이다.  
H1(대립가설)=클레임 접수 고객의 구매유형별 비율은 1회성구매형 10%, 실용적구매 30%,    
명품구매형 20%, 집중구매형 40%가 아니다.

In [11]:
Ob = X.values[1,:4]
Ob
Pr = np.array([0.1, 0.3, 0.2, 0.4])
n= X.values[1,4]
E=n*Pr
E

array([ 45.9, 137.7,  91.8, 183.6])

In [15]:
Ob/n*100

array([ 4.13943355, 26.14379085, 18.95424837, 50.76252723])

In [12]:
stats.chisquare(Ob,E)

Power_divergenceResult(statistic=31.582788671023955, pvalue=6.407672361002909e-07)

카이제곱 통계량은 31.6, 유의수준 p-value는 0.01이하로 출력되었다.  
즉, 클레임 접수 고객의 구매유형별 비율은 귀무가설을 따르지 않는다.

#### 독립성 검정
두개의 범주형 변수간에 연관성이 있는지, 독립적인지 통계적으로 판단  
H0(귀무가설)=클레임 접수 여부와 성별은 연관성이 없다.  
H1(대립가설)=클레임 접수 여부와 성별은 연관성이 있다.  
[성별] 0:남성, 1:여성

In [16]:
X=pd.crosstab(df.성별, df.클레임접수여부, margins=False)
X

클레임접수여부    0    1
성별               
0        139   10
1        402  449

In [17]:
stats.chi2_contingency(X)

(106.43790651350751,
 5.911607287197053e-25,
 1,
 array([[ 80.609,  68.391],
        [460.391, 390.609]]))

독립성 검정결과 카이제곱 통계량은 106.437, 유의수준 p-value는 0.01 이하로 도출되었다.   
따라서, 클레임접수여부와 성별은 연관성이 있다.  
결과에 따라, 남성과 여성 성별에 따라 다른 방식의 고객서비스를 제공함으로써 고객의 불만을 최소화 시킬 수 있는 방안을 모색해야 한다.

#### 동질성 검정    
서로 다른 모집단에서 독립적으로 추출한 표본들의 범주별 비율이 서로 동질적인 것인지 검정    
H0(귀무가설):청년층과 중장년층의 구매 유형비율은 동일하다.   
H1(대립가설):청년층과 중장년층의 구매 유형비율은 동일하지 않다.   
   

In [37]:
#고객_나이대 : 1(19세 이하), 2(20~24), 3(25~29), 4(30~34), 5(35~39), 
#              6(40~44), 7(45~49), 8(50~54), 9(55세 이상)
#구매유형 1:1회성구매형, 2:실용적구매형, 3: 명품구매형, 4:집중구매형
import warnings
warnings.filterwarnings('ignore')
df['고객연령대']=""

In [38]:
df.고객연령대[df.고객_나이대==1]="1",
df.고객연령대[df.고객_나이대==2]="1",
df.고객연령대[df.고객_나이대==3]="1",
df.고객연령대[df.고객_나이대==4]="1",
df.고객연령대[df.고객_나이대==5]="1",
df.고객연령대[df.고객_나이대==6]="2",
df.고객연령대[df.고객_나이대==7]="2",
df.고객연령대[df.고객_나이대==8]="2",
df.고객연령대[df.고객_나이대==9]="2"

In [42]:
df.고객연령대.value_counts()

1    609
2    391
Name: 고객연령대, dtype: int64

In [40]:
df1=df.loc[df.고객연령대=='1']
df2=df.loc[df.고객연령대=='2']

In [43]:
df1_sample=df1.sample(200, random_state=29)
df2_sample=df2.sample(200, random_state=29)
df3=df1_sample.append(df2_sample)

In [46]:
X=pd.crosstab(df3.고객연령대, df3.구매유형, margins=False)
stats.chi2_contingency(X)

(8.714970024248375,
 0.033330661685866646,
 3,
 array([[ 9., 66., 28., 97.],
        [ 9., 66., 28., 97.]]))

In [47]:
X

구매유형    1   2   3    4
고객연령대                 
1      10  67  18  105
2       8  65  38   89

동질성 검정 결과, 카이제곱 통계량은 8.71, p-value는 0.03으로 도출되었다.  
유의수준 95% 수준에서, A쇼핑고객들의 청년층과 중장년층의 구매유형 분포는 동질적이지 않다는 결론을 낼 수 있다.  
청년층의 경우 실용적 구매와 집중적 구매가 비중이 높으며, 중장년층의 경우 역시 실용적 구매와 집중적 구매의 비중이 높으나  
명품 구매형의 비중이 상대적으로 청년층 대비 높아 이에 대한 마케팅 전략 수립이 필요하다.